### PR TITLE
Feature: Relay getTxBySafeTxHash via interface, use client gateway

### LINKED ISSUE #104: Use new /multisig-transactions endpoint for txs.getBySafeTxHash
The endpoint we currently use is set to be deprecated in the near future. We need to check how many projects are using it and depending on that make the change.

### PR DESCRIPTION
**Pull request by mikheevm**
This PR:
- Closes #104 by adding a new interface message method
- Rework types by splitting them by domain/context

### CODE REVIEW

**Comment by mikheevm** [comment]
Depends on https://github.com/gnosis/safe-react/pull/2361

File: packages/safe-apps-sdk/src/txs/index.ts [hunk 1]
```diff
@@ -1,43 +1,42 @@
...
-    setTimeout(() => controller.abort(), 10000);
-
     try {
-      const res = await fetch(`${this.txServiceUrl}/transactions/${safeTxHash}`, options);
-      if (res.status !== 200) {
-        throw new Error(
-          "Failed to get the transaction. Either safeTxHash is incorrect or transaction hasn't been indexed by the service yet",
-        );
+      const response = await this.communicator.send<
+        'getTxBySafeTxHash',
```
Comment (katspaugh) [inline packages/safe-apps-sdk/src/txs/index.ts:24]: Can this be replaced by `METHODS.getTxBySafeTxHash`?

File: packages/safe-apps-sdk/src/txs/index.ts [hunk 2]
```diff
@@ -1,43 +1,42 @@
...
+      if (!response.success) {
+        throw new Error(response.error);
       }
 
-      const json = await res.json();
-
-      return json as TxServiceModel;
+      return response.data;
     } catch (err) {
       throw err;
```
Comment (katspaugh) [inline packages/safe-apps-sdk/src/txs/index.ts:35]: Not related to this PR but why catching it in the first place then?

File: packages/safe-apps-sdk/src/txs/index.ts [hunk 3]
```diff
@@ -47,7 +46,7 @@ class TXs {
       params,
     };
 
-    const response = await this.communicator.send<'sendTransactions', SendTransactionsArgs, SendTransactionsResponse>(
+    const response = await this.communicator.send<'sendTransactions', SendTransactionsParams, SendTransactionsResponse>(
```
Comment (katspaugh) [inline packages/safe-apps-sdk/src/txs/index.ts:49]: `sendTransactions` > `METHODS.sendTransactions` perhaps

File: packages/safe-apps-sdk/src/types/messaging.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import { METHODS } from '../communication/methods';
+import { SafeInfo } from './sdk';
+import { GatewayTransactionDetails, SafeBalances } from './gateway';
+
+export type Methods = keyof typeof METHODS;
```
Comment (katspaugh) [inline packages/safe-apps-sdk/src/types/messaging.ts:5]: Not related to the PR directly but... you wouldn't need this if METHODS was an enum.

**Review by katspaugh — CHANGES_REQUESTED** [review]
Looks good! Just some minor comments.

Is there any way to not show the dist files in PRs? Committing the dist after the review maybe?

File: packages/safe-apps-sdk/src/types/messaging.ts (hunk 4 above)
Comment (mikheevm) [reply packages/safe-apps-sdk/src/types/messaging.ts]: In general types for this sdk is messy and there's a room for big improvement I believe. The main thing I don't understand how to do properly is to have a connection between method/params/responses in both public and private methods. This is the reason for cumberson constructions such as:
```
  public send = <M extends Methods, P, R>(method: M, params: P): Promise<Response<R>> => {
  ```
  
If wish/have time to look at it and see if you have an idea, would really appreciate it 🙂

File: packages/safe-apps-sdk/src/txs/index.ts (hunk 2 above)
Comment (mikheevm) [reply packages/safe-apps-sdk/src/txs/index.ts:35]: I'm sorry, could you please elaborate? I see that the `(!response.success)`check can be moved to `communicator.send` method. Is this what you meant?

Comment (katspaugh) [reply packages/safe-apps-sdk/src/txs/index.ts:35]: I meant that it's catching and just re-throwing the same error.

**Review by katspaugh — APPROVED** [review]

**Pull request merged**